### ISSUE TITLE
JIT creates some unnecessary blocks/jumps

### ISSUE THREAD

**Issue by Sergio0694** [dotnet/runtime#45770]
### Description

Hi, I stumbled upon the JIT creating some extra unnecessary blocks in the codegen, as well as some seemingly weird jumps.

<details>
  <summary><b>Unnecessary block (C#, click to expand)</b></summary>
<br/>

```csharp
public class C
{
    public int x;

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public int AlignToStart1(int size)
    {
        return Align1(this.x, size, 16);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static int Align1(int offset, int size, int alignment)
    {
        if (((uint)offset / (uint)alignment) == ((uint)(offset + size) / (uint)alignment))
        {
            return offset;
        }

        return (offset + alignment - 1) & ~(alignment - 1);
    }
}
```
</details>

<details>
  <summary><b>Unnecessary block (asm, click to expand)</b></summary>
<br/>

```asm
; Assembly listing for method RefToArrayTest.C:AlignToStart1(int):int:this
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T01] (  3,  3   )     ref  ->  rcx         this class-hnd
;  V01 arg1         [V01,T02] (  3,  3   )     int  ->  rdx        
;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V03 tmp1         [V03,T03] (  3,  2   )     int  ->  rax         "Inline return value spill temp"
;  V04 tmp2         [V04,T00] (  5,  8   )     int  ->  rax         "Inlining Arg"
;
; Lcl frame size = 0

G_M5960_IG01:
						;; bbWeight=1    PerfScore 0.00
G_M5960_IG02:
       mov      eax, dword ptr [rcx+8]
       add      edx, eax
       shr      edx, 4
       mov      ecx, eax
       shr      ecx, 4
       cmp      edx, ecx
       jne      SHORT G_M5960_IG04
						;; bbWeight=1    PerfScore 7.75
G_M5960_IG03:
       jmp      SHORT G_M5960_IG05
						;; bbWeight=0.50 PerfScore 1.00
G_M5960_IG04:
       add      eax, 15
       and      eax, -16
						;; bbWeight=0.50 PerfScore 0.25
G_M5960_IG05:
       ret      
						;; bbWeight=1    PerfScore 1.00

; Total bytes of code 26, prolog size 0, PerfScore 12.60, instruction count 11 (MethodHash=2865e8b7) for method RefToArrayTest.C:AlignToStart1(int):int:this
; ============================================================
```
</details>

<details>
  <summary><b>Weird jump (C#, click to expand)</b></summary>
<br/>

```csharp
public class C
{
    public int x;

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public int AlignToStart2(int size)
    {
        return Align2(this.x, size, 16);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static int Align2(int offset, int size, int alignment)
    {
        if (((uint)offset / (uint)alignment) != ((uint)(offset + size) / (uint)alignment))
        {
            return (offset + alignment - 1) & ~(alignment - 1);
        }

        return offset;
    }
}
```
</details>

<details>
  <summary><b>Weird jump (asm, click to expand)</b></summary>
<br/>

```asm
; Assembly listing for method RefToArrayTest.C:AlignToStart2(int):int:this
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T01] (  3,  3   )     ref  ->  rcx         this class-hnd
;  V01 arg1         [V01,T02] (  3,  3   )     int  ->  rdx        
;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]   "OutgoingArgSpace"
;  V03 tmp1         [V03,T03] (  3,  2   )     int  ->  rax         "Inline return value spill temp"
;  V04 tmp2         [V04,T00] (  5,  8   )     int  ->  rax         "Inlining Arg"
;
; Lcl frame size = 0

G_M49803_IG01:
						;; bbWeight=1    PerfScore 0.00
G_M49803_IG02:
       mov      eax, dword ptr [rcx+8]
       add      edx, eax
       shr      edx, 4
       mov      ecx, eax
       shr      ecx, 4
       cmp      edx, ecx
       je       SHORT G_M49803_IG04
						;; bbWeight=1    PerfScore 7.75
G_M49803_IG03:
       add      eax, 15
       and      eax, -16
       jmp      SHORT G_M49803_IG04
						;; bbWeight=0.50 PerfScore 1.25
G_M49803_IG04:
       ret      
						;; bbWeight=1    PerfScore 1.00

; Total bytes of code 26, prolog size 0, PerfScore 12.60, instruction count 11 (MethodHash=0f443d74) for method RefToArrayTest.C:AlignToStart2(int):int:this
; ============================================================

```
</details>

In the first case, I would've expected that unconditional jump to the end and its corresponding block to be removed entirely, and the previous conditional jump to be just switched to `je G_M5960_IG03` (as in, what is `IG04` there). As for the second one, I'm referring to that last uncoditional ` jmp SHORT G_M49803_IG04` that just jumps to the instruction immediately after that.

### Configuration

- Codegen done with disasmo on a local checked build of .NET 5 from master (sometimes last week)

### Regression?

Can repro on .NET Core 3.1 too on sharplab, so not a regression.

**Comment by AndyAyersMS** [comment]
In the first example, we originally had a balanced `if-then-else` with a copy in the `then` block but LSRA manages to eliminate the copy by allocating both source and dest to `EAX`, creating an empty block. But we don't recognize this block is empty until emission time, and not until after we've already emitted the first jump.

| Before LSRA | After LSRA |
| :----: | :----: |
| ![image - 2020-12-08T130003 204](https://user-images.githubusercontent.com/10121823/101541002-a3eaa780-3955-11eb-8665-a15f3b1549da.png) | ![image - 2020-12-08T130116 613](https://user-images.githubusercontent.com/10121823/101541007-a64d0180-3955-11eb-89f3-56c2e79efcca.png) |

Second case is likely similar with just the order of the two intermediate blocks swapped.

We might consider revisiting block layout after LSRA, especially given its predilection for placing compensation blocks in odd places. 

Next step would be to write an assembly screener to see how often we end up with cases like these across a broader set of methods.

**Comment by AndyAyersMS** [comment]
Putting in future until we have more data on prevalence.

**Comment by Sergio0694** [comment]
That's super interesting!
The extra info and the two diagrams helped a lot to follow the explanation as well, thanks! 😊

**Comment by AndyAyersMS** [comment]
If you're curious about writing a screener to spot this sort of thing, you can build on what @kunalspathak has done [here](https://github.com/dotnet/jitutils/tree/master/src/AnalyzeAsm).

Though if it was me, I might use PMI instead of crossgen to generate the disassembly files.

**Comment by AndyAyersMS** [comment]
> [!NOTE]
> This comment was AI-assisted (GitHub Copilot CLI).

Re-checked this on current `main` (arm64, but the pattern is target-independent).

Both example methods still produce the suboptimal layout described originally — e.g. `AlignToStart2` inlined into a hot loop emits:

```
G_M27646_IG06: cmp w3, #2; bne G_M27646_IG08
G_M27646_IG07: mov w3, #42; b  G_M27646_IG09      ; jump over IG08 only
G_M27646_IG08: mov w3, #48                         ; falls through
G_M27646_IG09: ...
```

The 3-opt block layout (`fgSearchImprovedLayout`) and post-layout reverse-condition pass (`optOptimizePostLayout`) that were added since this was filed do not change the structure — once both branches are tiny one-store blocks before a merge, at least one of them needs an unconditional branch and layout alone can't fix that.

The JIT dump shows the if-conversion phase actually recognizes both candidates as `SELECT` patterns:

```
JTRUE block is BB02. Statement STMT00018 (Then) and STMT00017 (Else) can be expressed as SELECT
…
Skipping if-conversion inside loop (via weight)
```

…but rejects them at `ifconversion.cpp:564` because of the blanket "skip if-conversion in loops" heuristic. With `DOTNET_JitStressModeNames=STRESS_IF_CONVERSION_INNER_LOOPS DOTNET_JitStressModeNamesOnly=1`, both inlined branches collapse to clean `csel`s and the redundant blocks/jumps disappear:

```
cmp w0, #2
csel w1, w1, w2, ne
cmp w0, #2
csel w0, w2, w1, eq
```

So the underlying "extra blocks / weird jumps" pattern is really the symptom of the if-conversion-in-loops restriction, which is being actively tracked by #96380 (open, milestone 11.0.0, assigned to @dhartglassMSFT, with a plan to base the decision on SSA-derived dependency analysis).

Closing as a duplicate of #96380.

**Comment by AndyAyersMS** [comment]
Closing as duplicate of #96380 per analysis above.
